scylla-cdc/src/log_reader.rs

```diff
@@ -0,0 +1,213 @@
+use std::sync::Arc;
+
+use anyhow;
+use chrono::Duration;
+use core::time;
+use futures::future::RemoteHandle;
+use futures::stream::{FusedStream, FuturesUnordered, StreamExt};
+use futures::FutureExt;
+use scylla::Session;
+use std::cmp::max;
+
+use crate::cdc_types::GenerationTimestamp;
+use crate::consumer::ConsumerFactory;
+use crate::stream_generations::GenerationFetcher;
+use crate::stream_reader::StreamReader;
+
+pub struct CDCLogReader {
+    // Tells the worker to stop
+    // Usage of the "watch" channel will make it possible to change the the timestamp later,
+    // for example if somebody loses patience and wants to stop now not later
+    end_timestamp: tokio::sync::watch::Sender<Duration>,
+}
+
+impl CDCLogReader {
+    // Creates a reader and runs it immediately
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        session: &Arc<Session>,
+        keyspace: String,
+        table_name: String,
+        start_timestamp: Duration,
+        end_timestamp: Duration,
+        window_size: Duration,
+        safety_interval: Duration,
+        sleep_interval: time::Duration,
+        consumer_factory: Arc<dyn ConsumerFactory>,
+    ) -> (Self, RemoteHandle<anyhow::Result<()>>) {
+        let (end_timestamp_sender, end_timestamp_receiver) =
+            tokio::sync::watch::channel(end_timestamp);
+        let mut worker = CDCReaderWorker::new(
+            session,
+            keyspace,
+            table_name,
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            end_timestamp_receiver,
+            consumer_factory,
+        );
+        let (fut, handle) = async move { worker.run().await }.remote_handle();
+        tokio::task::spawn(fut);
+        let printer = CDCLogReader {
+            end_timestamp: end_timestamp_sender,
+        };
+        (printer, handle)
+    }
+
+    // Tell the worker to set the end timestamp and then stop
+    pub fn stop_at(&mut self, when: Duration) {
+        let _ = self.end_timestamp.send(when);
+    }
+
+    // Tell the worker to stop immediately
+    pub fn stop(&mut self) {
+        self.stop_at(Duration::min_value());
+    }
+}
+
+struct CDCReaderWorker {
+    session: Arc<Session>,
+    keyspace: String,
+    table_name: String,
+    start_timestamp: Duration,
+    end_timestamp: Duration,
+    sleep_interval: time::Duration,
+    window_size: Duration,
+    safety_interval: Duration,
+    readers: Vec<Arc<StreamReader>>,
+    end_timestamp_receiver: tokio::sync::watch::Receiver<Duration>,
+    consumer_factory: Arc<dyn ConsumerFactory>,
+}
+
+impl CDCReaderWorker {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        session: &Arc<Session>,
```

Comment:
Nitpick: it's better to accept by value if you are going to clone anyway. Now, there is a mandatory clone of the `Arc`, which involves a relatively expensive atomic operations, but if you accept by value then the caller will be able to choose if they want to move the `Arc` in or clone it.